### PR TITLE
operations with shortcuts and their focus effects

### PR DESCRIPTION
**Pull request by Levilian**
fix: #1612, #1227

### CODE REVIEW

**Comment by yipenghe** [comment]
Thanks for working on this!! Can you rebase this, I will test after

**Comment by Levilian** [comment]
@yipenghe

**Comment by Levilian** [comment]
fix: #1668

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 1]
```diff
@@ -342,6 +342,13 @@ class Worksheet extends React.Component {
 
     setFocus = (index, subIndex, shouldScroll = true) => {
         var info = this.state.ws.info;
+
+        // prevent multiple clicking from reseting the index
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js]:
```suggestion
        // prevent multiple clicking from resetting the index
```

File: frontend/src/components/worksheets/Worksheet/Worksheet.js [hunk 2]
```diff
@@ -975,36 +981,41 @@ class Worksheet extends React.Component {
...
                         // If the number of bundles increases then the focus should be on the new bundle.
                         // if the current focus is not on a table
-                        if (items[this.state.focusIndex] &&
-                            items[this.state.focusIndex].mode &&
-                            items[this.state.focusIndex].mode !== 'table_block') {
-                            this.setFocus(this.state.focusIndex + 1, 0);
+                        if (items[focus] &&
+                            items[focus].mode &&
+                            items[focus].mode !== 'table_block') {
+                            this.setFocus(focus >= 0 ? focus + 1 : 'end', 0);
```
Comment (epicfaace) [inline frontend/src/components/worksheets/Worksheet/Worksheet.js:1033]: Why add this conditional? Wouldn't passing in -1 to the first argument of `setFocus` have the same affect as passing in "end"?

File: frontend/src/components/worksheets/items/MarkdownItem.js [hunk 3]
```diff
@@ -52,34 +54,38 @@ class MarkdownItem extends React.Component {
...
-        // Delete the line
-        Mousetrap.bind(
-            ['backspace', 'del'],
-            function() {
-                if (this.props.focusIndex >= 0) {
-                    this.deleteItem();
-                }
-            }.bind(this),
-        );
+        // var focusElem = document.querySelector('.type-markup.focused');
```
Comment (epicfaace) [inline frontend/src/components/worksheets/items/MarkdownItem.js]: Can you remove these commented out lines?

File: frontend/src/components/worksheets/items/MarkdownItem.js [hunk 4]
```diff
@@ -104,9 +109,9 @@ class MarkdownItem extends React.Component {
     };
 
     render() {
-        this.capture_keys(); // each item capture keys are handled dynamically after this call
+        // this.capture_keys();
```
Comment (epicfaace) [inline frontend/src/components/worksheets/items/MarkdownItem.js]: remove this

File: frontend/src/components/worksheets/Worksheet/Worksheet.js (hunk 2 above)
Comment (Levilian) [reply frontend/src/components/worksheets/Worksheet/Worksheet.js:1033]: Nope. One would deselect the element whereas passing in "end" would select the last element.

**Pull request merged**